symengine/sets.cpp

```diff
@@ -447,6 +529,35 @@ RCP<const Set> FiniteSet::set_intersection(const RCP<const Set> &o) const
         }
         return finiteset(container);
     }
+    if (is_a<Reals>(*o)) {
+        set_basic kept_reals;
+        set_basic others;
+        for (const auto &elem : container_) {
+            if (is_a_Number(*elem)) {
+                if (!down_cast<const Number &>(*elem).is_complex()) {
+                    kept_reals.insert(elem);
+                }
+            } else {
+                others.insert(elem);
+            }
+        }
+        if (kept_reals.empty()) {
+            if (others.empty()) {
+                return emptyset();
+            } else {
+                return SymEngine::set_intersection(
+                    {reals(), finiteset(others)});
+            }
+        } else {
+            if (others.empty()) {
+                return finiteset(kept_reals);
+            } else {
+                others.insert(kept_reals.begin(), kept_reals.end());
+                return SymEngine::set_intersection(
+                    {reals(), finiteset(others)});
```

Comment:
```suggestion
                return set_union(finiteset(kept_reals), set_intersection(
                    {reals(), finiteset(others)}));
```

Reply:
Above is another way to return it. Do you know what sympy does?

Reply:
I tested:

```python
import sympy

x = sympy.Symbol('x')
finite = sympy.FiniteSet(sympy.Integer(1), sympy.Integer(2), sympy.I, x)
sympy.Reals.intersection(finite)
```

And got

```
Intersection(FiniteSet(1, 2, x), Reals)
```

Should be in line with my suggestion

Reply:
Great. We'll go with your version